### DISCUSSION TITLE
YOLOv5 Upstream Entry Point Differences

### DISCUSSION THREAD

**Discussion by SIR-unit** [fcakyon/yolov5-pip#225] [Q&A]
I noticed there are some differences with upstream.  Specifically, this code has parse_opt() moved into main() for detect.py and val.py.  I would like to be able to do something like so in my code:
```
assert mode in ('train', 'val', 'detect')
module = import_module(f'yolov5.{mode}')
opt = module.parse_opt()
opt = my_modify_opt_func(opt)
main(opt)
```
Simply calling run() does not work because, in train.py, it calls parse_opt(True), which causes misspelled arguments to silently fail, and in detect.py and val.py, run() doesn't call parse_opt() at all.  More importantly, main() has other code which is useful when running from cli, so I really want to be able to call main() with other options; this is especially true for val.py, whose main() does things other than check dependencies and call run()

I understand that you might not want to break codes that rely on this difference, but If I make a pull request with a backwards-compatible change which addresses this, would it be accepted?  Specifically, I would change the first part of main() in detect.py and val.py to
```
def main(opt=None):
    if opt is not None:
        opt = parse_opt()
```

**Comment by fcakyon** [comment]
@SIR-unit if you make backwards compatible changes, it would be ok by me 👍

**Reply by SIR-unit** [reply]
Awesome.  I've made pull requests for val.py and detect.py.

Also, I've been having trouble with an unrelated bug in val.py when trying to evaluate with normal precision on GPU.  I've also made pull request to address that.
